reana_job_controller/compute4punch_job_manager.py

```diff
@@ -0,0 +1,366 @@
+            )
+        )
+
+    @classmethod
+    def _download_output_directory(
+        cls, sftp_client: SFTPClient, remote_dir: str, local_dir: str
+    ) -> None:
+        """
+        Download output directory and content to a local directory.
+
+        :param sftp_client: SFTP client to use for downloading
+        :type sftp_client: SFTPClient
+        :param remote_dir: Remote directory to download
+        :type remote_dir: str
+        :param local_dir: Local destination directory
+        :type local_dir: str
+        """
+        skipped_output_files = (
+            "cluster.log",
+            "_condor_stdout",
+            "_condor_stderr",
+            "run.sh",
+            "submit.jdl",
+            "tmp",
+            "var_tmp",
+        )  # skip intermediate and temporary htcondor files
+        os.path.exists(local_dir) or os.makedirs(local_dir)
+        for item in filter(
+            lambda x: x.filename not in skipped_output_files,
+            sftp_client.listdir_attr(remote_dir),
+        ):
+            remote_path = os.path.join(remote_dir, item.filename)
+            local_path = os.path.join(local_dir, item.filename)
+            if S_ISDIR(item.st_mode):
+                cls._download_output_directory(sftp_client, remote_path, local_path)
+            else:
+                sftp_client.get(remote_path, local_path)
+
+    @staticmethod
+    def _encode_cmd(cmd: str) -> str:
+        """
+        Encode base64 cmd.
+
+        :param cmd: Command to encode
+        :type cmd: str
+        """
+        encoded_cmd = base64.b64encode(cmd.encode("utf-8")).decode("utf-8")
+        return f"echo {encoded_cmd} | base64 -d | bash"
+
+    def _get_inputs(self) -> Iterable:
+        """
+        Collect all input files in the local REANA workspace.
+        """
+        skipped_input_files = (".job.ad", ".machine.ad", ".chirp.config")
+        return filter(
+            lambda x: x not in skipped_input_files,
+            walk(
+                workspace=self.workflow_workspace,
+            ),
+        )
+
+    def _upload_job_inputs(self) -> None:
+        """
+        Upload job inputs to Compute4PUNCH
+        """
+        sftp_client = self.c4p_connection.ssh_client.open_sftp()
+        sftp_client.chdir(self.c4p_rel_workspace_path)
+
+        try:
+            for job_input in self._get_inputs():
+                if is_directory(self.workflow_workspace, job_input):
+                    try:  # check if directory already exists
+                        sftp_client.stat(job_input)
+                    except FileNotFoundError:
+                        sftp_client.mkdir(job_input)
+                    finally:
+                        continue
+                else:
+                    local_path = os.path.join(self.workflow_workspace, job_input)
+                    remote_path = os.path.join(self.c4p_abs_workspace_path, job_input)
+                    sftp_client.put(local_path, remote_path)
```

Comment:
As for the download counterpart, are we sure here that the parent directories are created before the child directories/files?

Reply:
I think we are fine here as well. You are using `os.fwalk` in https://github.com/reanahub/reana-commons/blob/b4a81d58d019ba3c4c155ac6ff0cc972fb162682/reana_commons/workspace.py#L169-L174, which `topdown: bool = True`. According to the [documentation](https://docs.python.org/3/library/os.html#os.walk) it means:

> If optional argument topdown is True or not specified, the triple for a directory is generated before the triples for any of its subdirectories (directories are generated top-down). If topdown is False, the triple for a directory is generated after the triples for all of its subdirectories (directories are generated bottom-up).

As far as I understand, the sub-directories are always created before the actual content is written.